src/encoder.c

```diff
@@ -88,69 +352,265 @@ static int p11prov_rsa_print_public_key(const OSSL_PARAM *params, void *bio)
+    type_name = p11prov_key_type_name(key);
+    if (!type_name) {
+        P11PROV_raise(ctx->provctx, CKR_GENERAL_ERROR, "Unknown Key Type");
         return RET_OSSL_ERR;
     }
-    class = p11prov_obj_get_class(key);
+
+    P11PROV_debug("%s Text Encoder", type_name);
 
     out = BIO_new_from_core_bio(p11prov_ctx_get_libctx(ctx->provctx), cbio);
     if (!out) {
         P11PROV_raise(ctx->provctx, CKR_GENERAL_ERROR, "Failed to init BIO");
         return RET_OSSL_ERR;
     }
 
+    class = p11prov_obj_get_class(key);
     keysize = p11prov_obj_get_key_bit_size(key);
 
     if (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) {
         if (class != CKO_PRIVATE_KEY) {
             BIO_printf(out, "[Error: Invalid key data]\n");
             goto done;
         }
-        BIO_printf(out, "PKCS11 RSA Private Key (%lu bits)\n", keysize);
+        BIO_printf(out, "PKCS11 %s Private Key (%lu bits)\n", type_name,
+                   keysize);
         BIO_printf(out, "[Can't export and print private key data]\n");
     }
 
     if (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) {
-        if (class != CKO_PUBLIC_KEY) {
-            P11PROV_OBJ *assoc;
+        CK_ATTRIBUTE *pkeyinfo = NULL;
+        P11PROV_OBJ *public = NULL;
+        bool free_public = false;
 
-            assoc = p11prov_obj_find_associated(key, CKO_PUBLIC_KEY);
-            if (!assoc) {
-                BIO_printf(out, "[Error: Failed to source public key data]\n");
-                goto done;
-            }
+        BIO_printf(out, "PKCS11 %s Public Key (%lu bits)\n", type_name,
+                   keysize);
 
-            /* replace key before printing the rest */
-            key = assoc;
+        if (class == CKO_PUBLIC_KEY) {
+            pkeyinfo = p11prov_obj_get_attr(key, CKA_PUBLIC_KEY_INFO);
+            if (pkeyinfo) {
+                if (pkeyinfo->ulValueLen == 0) {
+                    pkeyinfo = NULL;
+                }
+            }
+            if (!pkeyinfo) {
+                public = key;
+            }
+        } else {
+            public = p11prov_obj_get_associated(key);
+            if (public && p11prov_obj_get_class(public) == CKO_PUBLIC_KEY) {
+                pkeyinfo = p11prov_obj_get_attr(public, CKA_PUBLIC_KEY_INFO);
+                if (pkeyinfo && pkeyinfo->ulValueLen > 0) {
+                    public = NULL;
+                } else {
+                    pkeyinfo = NULL;
+                }
+            }
+            if (!public && !pkeyinfo) {
+                pkeyinfo = p11prov_obj_get_attr(key, CKA_PUBLIC_KEY_INFO);
+                if (pkeyinfo && pkeyinfo->ulValueLen == 0) {
+                    pkeyinfo = NULL;
+                }
+            }
+            if (!public && !pkeyinfo) {
+                /* Finally try the most expensive option */
+                public = p11prov_obj_find_associated(key, CKO_PUBLIC_KEY);
+                free_public = true;
+            }
```

Comment:
This order sounds to me odd -- first we try `CKA_PUBLIC_KEY_INFO`  from associated public key object, then the `CKA_PUBLIC_KEY_INFO` from the private key object and then we search for the associated object.
Should not the first two be swapped?